synapse/replication/tcp/handler.py

```diff
@@ -122,17 +129,42 @@ def __init__(self, hs):
         # outgoing replication commands to.)
         self._connections = []  # type: List[AbstractConnection]
 
-        # For each connection, the incoming stream names that are coming from
-        # that connection.
-        self._streams_by_connection = {}  # type: Dict[AbstractConnection, Set[str]]
-
         LaterGauge(
             "synapse_replication_tcp_resource_total_connections",
             "",
             [],
             lambda: len(self._connections),
         )
 
+        # When POSITION or RDATA commands arrive, we stick them in a queue and process
+        # them in order in a separate background process.
+
+        # the streams which are currently being processed by _unsafe_process_stream
+        self._processing_streams = set()  # type: Set[str]
+
+        # for each stream, a queue of commands that are awaiting processing, and the
+        # connection that they arrived on.
+        self._command_queues_by_stream = {
+            stream_name: Deque[
+                Tuple[Union[RdataCommand, PositionCommand], AbstractConnection]
+            ]()
```

Comment:
FWIW, from PEP 484:

> It is not recommended to use the subscripted class (e.g. `Node[int]`) directly in an expression -- using a type alias (e.g. `IntNode = Node[int]`) instead is preferred. (First, creating the subscripted class, e.g. `Node[int]`, has a runtime cost. Second, using a type alias is more readable.)

Since this is instantiating the type a few times, and the type is going across 3 lines, probably worth aliasing it?

Reply:
great suggestion, thanks.